src/optimal_image/compress/image.rs

```diff
@@ -56,6 +70,89 @@ impl<'a> VipsImage<'a> {
         }
     }
 
+    pub fn from_buffer(
+        buf: *const c_void,
+        len: usize,
+    ) -> Result<VipsImage<'a>, Box<Error>> {
+        let c =
+            unsafe { vips::vips_image_new_from_buffer(buf, len, NULL_LIST) };
+        result(c)
+    }
+
+    pub fn from_memory(
+        data: *const c_void,
+        size: usize,
+        width: i32,
+        height: i32,
+        bands: i32,
+    ) -> Result<VipsImage<'a>, Box<Error>> {
+        let c = unsafe {
+            vips::vips_image_new_from_memory_copy(
+                data,
+                size,
+                width,
+                height,
+                bands,
+                vips::VipsBandFormat::VIPS_FORMAT_UCHAR,
+            )
+        };
+        result(c)
+    }
+
+    pub fn from_image_data(
+        img: &ImageData,
+    ) -> Result<VipsImage<'a>, Box<Error>> {
+        let mut byte_data: Vec<u8> =
+            img.pixels().fold(Vec::new(), |mut data: Vec<u8>, pixel| {
+                let rgb::RGBA { r, g, b, a } = pixel;
+                data.push((r * 255.0) as u8);
+                data.push((g * 255.0) as u8);
+                data.push((b * 255.0) as u8);
+                data.push((a * 255.0) as u8);
+                data
+            });
+
+        let byte_data_ptr = byte_data.as_mut_ptr() as *const c_void;
+        let bands: i32 = 4; // RGBA (4 bands)
+        let size = byte_data.len() as usize;
+        let width = img.width() as i32;
+        let height = img.height() as i32;
+
+        let c =
+            VipsImage::from_memory(byte_data_ptr, size, width, height, bands)?;
+        result(c.img)
+    }
+
+    pub fn to_image_data(&self) -> Result<ImageData, Box<Error>> {
+        let vector: Vec<u8> = unsafe {
+            let mut result_size: usize = 0;
+            let memory: *mut u8 = vips::vips_image_write_to_memory(
+                self.img,
+                &mut result_size as *mut usize,
+            ) as *mut u8;
+            let slice = slice::from_raw_parts_mut(memory, result_size);
+            let vec = slice.to_vec();
+            // manually free the memory
+            libc::free(memory as *mut libc::c_void);
```

Comment:
This means we are converting the `VipsImage<T>` to ImageData and then freeing up the VipsImage memory. Is it possible that any successive function calls on `VipsImage` could panic? e.g.

```
VipsImage::to_image_data();
VipsImage::jpegsave(path, 70); // <------ Could this panic?
```